### PR TITLE
Properly adjust target time when sub-cycling with negative times

### LINKED ISSUE #15766: Sub_cycling sync time not properly adjusting with negative times
## Bug Description

Until #14865 is fixed, a workaround is to run a pseudo-transient from `start_time = some_negative_time` to `end_time = 0` and to restart the actual transient from that solution.

However, when such a pseudo-transient is performed with a main app driving a sub-app and `sub_cycling` is used, the sub-app may become out of sync with the main app. This is because the sub-app picks its own time step with `sub_cycling` and is supposed to make sure that the chosen `dt` does not put it after the `target_time`.

Unfortunately, the current check is:

```
// Adjust to a target time if set
  if (_target_time > 0 && _time + dt_cur + _timestep_tolerance >= _target_time)
  {
    dt_cur = _target_time - _time;
    _at_sync_point = true;

    diag << "Limiting dt for target time: " << std::setw(9) << std::setprecision(6)
         << std::setfill('0') << std::showpoint << std::left << _next_interval_output_time
         << " dt: " << std::setw(9) << std::setprecision(6) << std::setfill('0') << std::showpoint
         << std::left << dt_cur << std::endl;
  }
```

The check `_target_time > 0` is not pertinent with negative times. I am not 100% sure why there is that condition in this if-statement but it seems to me that it is because we have `target_time` set by default to -1:

```
_target_time(declareRecoverableData<Real>("target_time", -1)),
```

If instead, I apply the following patch:

```
diff --git a/framework/src/executioners/Transient.C b/framework/src/executioners/Transient.C
index c96fa7c..c4a36d0 100644
--- a/framework/src/executioners/Transient.C
+++ b/framework/src/executioners/Transient.C
@@ -155,7 +155,7 @@ Transient::Transient(const InputParameters & parameters)
     _time_interval(declareRecoverableData<bool>("time_interval", false)),
     _start_time(getParam<Real>("start_time")),
     _timestep_tolerance(getParam<Real>("timestep_tolerance")),
-    _target_time(declareRecoverableData<Real>("target_time", -1)),
+    _target_time(declareRecoverableData<Real>("target_time", -std::numeric_limits<Real>::max())),
     _use_multiapp_dt(getParam<bool>("use_multiapp_dt")),
     _solution_change_norm(declareRecoverableData<Real>("solution_change_norm", 0.0)),
     _sln_diff(_nl.addVector("sln_diff", false, PARALLEL)),
@@ -531,7 +531,7 @@ Transient::computeConstrainedDT()
   }
 
   // Adjust to a target time if set
-  if (_target_time > 0 && _time + dt_cur + _timestep_tolerance >= _target_time)
+  if (_target_time > -std::numeric_limits<Real>::max() + _timestep_tolerance && _time + dt_cur + _timestep_tolerance >= _target_time)
   {
     dt_cur = _target_time - _time;
     _at_sync_point = true;
```

Everything behaves as expected (in the same way as it does with positive times).

## Steps to Reproduce

Currently, I do not have any simple example reproducing the error that I can share.

## Impact

This is problematic because the sub-apps become out-of-sync.

### PR DESCRIPTION
**Pull request by vincentlaboure**
Close #15766

### CODE REVIEW

**Comment by vincentlaboure** [comment]
This is ready for review.

**Comment by vincentlaboure** [comment]
Added fix+test to address #15802

**Comment by loganharbour** [comment]
Please push a change to this to forcefully re-trigger testing (full testing is now available).

You can do the following:
```
git commit --amend
```
and save the amend message that comes up without changing it (by typing `:wq` if in vim). This will forcefully change the commit hash of the last commit, which will trigger a change. Then forcefully push:
```
git push --force
```

**Comment by tophmatthews** [comment]
`multiapps/transient_multiapp.test_negative_time: Tester failed, reason: MISSING GOLD FILE` @vincentlaboure

**Comment by vincentlaboure** [comment]
Thanks @tophmatthews! For some reason, I try to click on the details but the page doesn't download... I'll fix it.

**Comment by vincentlaboure** [comment]
Seems to me that the recover test is showing yet another bug

**Comment by vincentlaboure** [comment]
There are two potential ways to fix the issue:

```
diff --git a/framework/src/base/MooseApp.C b/framework/src/base/MooseApp.C
index 7800fac4c3..81061f4703 100644
--- a/framework/src/base/MooseApp.C
+++ b/framework/src/base/MooseApp.C
@@ -292,7 +292,7 @@ MooseApp::MooseApp(InputParameters parameters)
     _file_base_set_by_user(false),
     _output_position_set(false),
     _start_time_set(false),
-    _start_time(0.0),
+    _start_time(-std::numeric_limits<Real>::max()),
     _global_time_offset(0.0),
     _output_warehouse(*this),
     _input_parameter_warehouse(new InputParameterWarehouse()),
```

which sets the default to negative values.
The problem is that in the `Transient` executioner, we have the following:

```
  if (_app.hasStartTime())
    _start_time = _app.getStartTime();
  else if (parameters.isParamSetByUser("start_time") && !_app.isRecovering())
    _app.setStartTime(_start_time);
```

so if we are recovering, the `_start_time` in `Transient` and `MooseApp` are not synchronized if none of these two conditions is satisfied. Another potential fix is then to remove `!_app.isRecovering()` from the `else if` statement:

```
diff --git a/framework/src/executioners/Transient.C b/framework/src/executioners/Transient.C
index d027607dd7..de5099515a 100644
--- a/framework/src/executioners/Transient.C
+++ b/framework/src/executioners/Transient.C
@@ -184,12 +184,14 @@ Transient::Transient(const InputParameters & parameters)
 
   // Either a start_time has been forced on us, or we want to tell the App about what our start time
   // is (in case anyone else is interested.
   if (_app.hasStartTime())
     _start_time = _app.getStartTime();
-  else if (parameters.isParamSetByUser("start_time") && !_app.isRecovering())
+  else if (parameters.isParamSetByUser("start_time"))
     _app.setStartTime(_start_time);
 
   _time = _time_old = _start_time;
```

I went with the first fix but let me know if you prefer the second one. Both make the `multiapps/transient_multiapp.test_negative_time` test pass.

**Comment by vincentlaboure** [comment]
Should be ready for review

**Comment by vincentlaboure** [comment]
Any chance someone can review this? I'm expecting it to be a fairly quick review and some of our customers will need that. Thanks!

**Comment by permcody** [comment]
@fdkong or @aeslaughter - please take a look at this. Let's make sure the NRC collaborators have what they need. Thanks!

File: framework/src/executioners/Transient.C [hunk 1]
```diff
@@ -536,7 +536,8 @@ Transient::computeConstrainedDT()
   }
 
   // Adjust to a target time if set
```
Comment (aeslaughter) [inline framework/src/executioners/Transient.C:538]: Please improve this to explain this rather complicated if statement.

**Review by aeslaughter — CHANGES_REQUESTED** [review]

File: framework/src/executioners/Transient.C (hunk 1 above)
Comment (vincentlaboure) [reply framework/src/executioners/Transient.C]: Changed to: `// If a target time is set and the current dt would exceed it, limit dt to match the target`

File: framework/src/executioners/Transient.C [hunk 2]
```diff
@@ -160,7 +160,7 @@ Transient::Transient(const InputParameters & parameters)
     _time_interval(declareRecoverableData<bool>("time_interval", false)),
     _start_time(getParam<Real>("start_time")),
     _timestep_tolerance(getParam<Real>("timestep_tolerance")),
-    _target_time(declareRecoverableData<Real>("target_time", -1)),
+    _target_time(declareRecoverableData<Real>("target_time", -std::numeric_limits<Real>::max())),
```
Comment (fdkong) [inline framework/src/executioners/Transient.C:163]: If we do not set start_time in the input file, I will have a negative `start_time`?

**Review by fdkong — APPROVED** [review]
It looks good for me

File: framework/src/executioners/Transient.C (hunk 2 above)
Comment (vincentlaboure) [reply framework/src/executioners/Transient.C:163]: This PR does not change the behavior w.r.t `start_time`. For the `_target_time`, it was currently assumed that if it is negative, then it has not been set. This logic does not work if we are starting the simulation at negative times, in which case the `_target_time` could have been set but still be negative.

**Comment by vincentlaboure** [comment]
@aeslaughter Should be good to go if you don't have any more comments.

**Review by aeslaughter — APPROVED** [review]

**Comment by vincentlaboure** [comment]
Thanks @permcody, @aeslaughter and @fdkong!

**Pull request merged**